Errors Concept

colinleach opened the pull request:
Please don't spend much time on this yet, it still needs some proofreading and polishing.

I'm aiming to port `javascript/factory-sensors` to pair with it. That isn't yet well advanced.

colinleach commented:
I've hacked this about a bit since the original submission. Meanwhile, `factory-sensors` is taking shape, now I've accepted that the JS original is a loose inspiration rather than something I can port directly.

I've beefed up the discussion on logging (`@info`, etc), and  that will be part of a task in the exercise. Changing the concept to "Errors and Logging" probably isn't worthwhile.

colinleach commented:
The Julia manual seemed to suggest that testing for log entries is simple and well-supported. I'm not having much success, so I may back out of most of what I was attempting.

depial commented:
> I'm not having much success, so I may back out of most of what I was attempting.

I'm not familiar with logging, so I'm afraid I can't be of much help.

I'll give you some more time to decide on what you would like to include before I review, but I'll get to this tomorrow regardless since my availability may drop off starting Wednesday.

colinleach commented:
I'm fine with the concept being reviewed. My problems are limited to Task 3 of `factory-sensors`: a nice-to-have, but maybe not worth this much aggravation.

I'll try to PR an exercise by the end of today, but it may be a draft-for-comment.

depial commented on concepts/errors/about.md:
````diff
@@ -0,0 +1,208 @@
+
+At any stage, the error can be intercepted with a [`try...catch`][try-catch] block which attempts to handle it.
+
+```julia
````
Should this be `julia-repl`?

depial commented on concepts/errors/about.md:
````diff
@@ -0,0 +1,208 @@
+0.0 + 3.141592653589793im  # success
+```
+
+In the example above, `log(n)` needs `n` to be either a positive real value, or any complex value.
+The `try ... catch` traps problems with negative real values, returning the correct complex answer `iπ` in mathematical notation.
+
+If you supply, for example, a string argument, there is no recovery except asking the user to correct it.
+
+As a final catch-all, we added [`rethrow()`][rethrow] for anything which is neither `DomainError` nor `MethodError`.
+
+## Logging
+
+Note that the `error()` function, discussed above, should not be confused with the [`@error`][error_macro] macro.
+
+The function generates an exception, which will be passed up the call stack unless caught.
+
+The `@error` macro, along with its [`@debug`][debug_macro], [`@info`][info_macro] and [`@warn`][warn_macro] counterparts, is part of the [`Logging`][logging] module, and intended to generate informative messages without altering program flow.
+
+Output goes to the terminal by default (color-coded by severity), though in a real application there are many other possibilities.
+
+```julia-repl
+julia> @warn "Something looks not quite right"
+┌ Warning: Something looks not quite right
+└ @ Main REPL[55]:1
+
+julia> @error "Panic!"
+┌ Error: Panic!
+└ @ Main REPL[56]:1
+```
````
Is there a reason the formatting makes all of this text blue? I can't see anything wrong with the previous code block, but it seems a bit odd to me.

Also, you use the `@warn` and `@info` macros in the previous example. Maybe it could be good to mention them somewhere in here?

colinleach replied:
> Is there a reason the formatting makes all of this text blue?

I'm not seeing it as blue, so I don't know what's happening.

> The hazards of moving the order of sections around. I've changed it, but it may still be non-ideal.

depial commented on concepts/errors/about.md:
````diff
@@ -0,0 +1,208 @@
+ERROR: AssertionError: n must be odd
+```
+
+## `try`...`catch`
````
It might be nice to mention somewhere that this is rather poor in terms of performance in comparison to using `if-else` logic. I'm not sure if that could be worked in smoothly or not though.

depial replied:
I forgot to mention... Do you think it could be too much information to include something on `else` and `finally`? I think they are interesting features, but I wouldn't want `about.md` getting too bulky.

depial reviewed (approved):
Just small things. I think this summed up the ideas nicely. 

Aside: I'm debating whether I would have put `custom errors` before `throw`, since it seems to follow from `standard error types`, but I also see the logic in the current order. I could go either way, so I just thought I'd throw that out there.

colinleach commented:
I'll merge for now, but yell at me if I haven't addressed all your comments. I want to see how this looks on the website, given confusion about blue text.

depial commented:
It looks like the blue text is gone. Maybe changing that to `julia-repl` made the difference?

depial commented:
> I'll merge for now, but yell at me if I haven't addressed all your comments.

The only thing I would point at is that I feel we may want to mention [`else`](https://docs.julialang.org/en/v1/manual/control-flow/#else-Clauses) clauses if we also mention `finally` clauses.

colinleach commented:
> we may want to mention else clauses if we also mention finally clauses

I'll submit another PR a bit later. I'm close to getting a draft of `factory-sensors` out, so I'd better finish that before I lose the mental thread.
